src/main/java/co/cask/publisher/GooglePublisher.java

```diff
@@ -0,0 +1,151 @@
+import co.cask.cdap.api.annotation.Macro;
+import co.cask.cdap.api.annotation.Name;
+import co.cask.cdap.api.annotation.Plugin;
+import co.cask.cdap.api.data.batch.Output;
+import co.cask.cdap.api.data.format.StructuredRecord;
+import co.cask.cdap.api.dataset.lib.KeyValue;
+import co.cask.cdap.etl.api.Emitter;
+import co.cask.cdap.etl.api.PipelineConfigurer;
+import co.cask.cdap.etl.api.batch.BatchSink;
+import co.cask.cdap.etl.api.batch.BatchSinkContext;
+import co.cask.cdap.format.StructuredRecordStringConverter;
+import co.cask.hydrator.common.batch.sink.SinkOutputFormatProvider;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.io.NullWritable;
+import org.apache.hadoop.io.Text;
+
+import javax.annotation.Nullable;
+
+/**
+ * Google Publisher sink to write to a Google PubSub topic.
+ */
+@Plugin(type = BatchSink.PLUGIN_TYPE)
+@Name("GooglePublisher")
+@Description("Batch Sink to write to Google Cloud Pub/Sub topic. Cloud Pub/Sub brings the scalability, " +
+  "flexibility, and reliability of enterprise message-oriented middleware to the cloud. By providing many-to-many, " +
+  "asynchronous messaging that decouples senders and receivers, it allows for secure and highly available " +
+  "communication between independently written applications")
+public class GooglePublisher extends BatchSink<StructuredRecord, NullWritable, Text> {
+  private final Config config;
+
+  @SuppressWarnings("unused")
+  public GooglePublisher(Config config) {
+    this.config = config;
+  }
+
+  @Override
+  public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
+    super.configurePipeline(pipelineConfigurer);
+    config.validate();
+  }
+
+  @Override
+  public void prepareRun(BatchSinkContext context) throws Exception {
+    config.validate();
+    Configuration configuration = new Configuration();
+    PubSubOutputFormat.configure(configuration, config);
+    context.addOutput(Output.of(config.referenceName,
+                                new SinkOutputFormatProvider(PubSubOutputFormat.class, configuration)));
+  }
+
+  @Override
+  public void transform(StructuredRecord input, Emitter<KeyValue<NullWritable, Text>> emitter) throws Exception {
+    String body = StructuredRecordStringConverter.toJsonString(input);
+    emitter.emit(new KeyValue<>(NullWritable.get(), new Text(body)));
+  }
+
+  /**
+   * PubSub Publisher config
+   */
+  public static class Config extends GCPReferenceConfig {
+    @Description("Cloud Pub/Sub topic to publish the records")
+    @Macro
+    public String topic;
+
+    // batching options
+    @Description("Element count threshold used for batching, default : 100")
+    @Macro
+    @Nullable
+    public Long messageCountBatchSize;
+
+    @Description("Request byte threshold used for batching, default : 1kb")
+    @Macro
+    @Nullable
+    public Long requestThresholdKB;
+
+    @Description("Delay threshold in millis from last publish for publishing the next batch, default : 1 ms")
```

Comment:
batch. The default value is 1 millisecond.

Reply:
I don't understand what this is. Does it mean batches are not published within 1ms of each other?

Reply:
a batch will be published either if message count exceeds threshold, total size in bytes exceeds threshold or the time since last publish exceeded the delay threshold.